core/common/src/main/java/alluxio/collections/LockCache.java

```diff
@@ -174,51 +203,24 @@ public ReentrantReadWriteLock getRawReadWriteLock(K key) {
 
   private ValNode getValNode(K key) {
     Preconditions.checkNotNull(key, "key can not be null");
-    ValNode cacheEntry;
-    while (true) {
-      // repeat until we get a cacheEntry that is not in the process of being removed.
-      cacheEntry = mCache.compute(key, (k, v) -> {
-        if (v != null) {
-          // we want to maintain the invariant that only one thread is removing from the map.
-          // Returning anything other than v would break that invariant.
-          v.mRefCount.incrementAndGet();
-          v.mIsAccessed = true;
-          return v;
-        }
-        // new cache entry
-        if (mCache.size() >= mHardLimit) {
-          return null;
-        } else {
-          return new ValNode(mDefaultLoader.apply(k));
-        }
-      });
-
-      if (cacheEntry == null) {
-        // cache is at hard limit
-        try {
-          if (System.currentTimeMillis() - mLastSizeWarningTime > 60000) {
-            LOG.warn("Cache at hard limit, cache size = " + mCache.size()
-                + " softLimit = " + mSoftLimit + " hardLimit = " + mHardLimit);
-            mLastSizeWarningTime = System.currentTimeMillis();
-          }
-          Thread.sleep(5);
-          evictIfOverLimit();
-        } catch (InterruptedException e) {
-          throw new RuntimeException(e);
-        }
-        continue;
+    ValNode cacheEntry = mCache.compute(key, (k, v) -> {
+      if (v != null && v.mRefCount.incrementAndGet() > 0) {
+        // If the entry is to be removed, ref count will be INT_MIN, so incrementAndGet will < 0.
+        v.mIsAccessed = true;
+        return v;
       }
-      evictIfOverLimit();
-      if (cacheEntry.mRefCount.get() > 0) {
-        return cacheEntry;
-      } else {
+      return new ValNode(mDefaultLoader.apply(k));
+    });
+    if (mCache.size() > mSoftLimit) {
```

Comment:
we probably want to redefine what soft and hard limits mean. I think this should signal the evictor if the hard limit is exceeded, to avoid needing to evict constantly around soft limit.

Reply:
softlimit is 0.9 * hardlimit, reaching softlimit is like a warning signal that hardlimit will be reached soon, so doing eviction around softlimit makes sense.

Reply:
I think hard and soft limit should behave like high and low watermarks. When greater than high watermark, evict down to the lower watermark.